### PR TITLE
Get privacy groups

### PR DESCRIPTION
**Pull request by mengyangwang**
Summary
---------

add new sync_privacy_groups function in fecher.py(class ThreatExchangeConfig) to fetch privacy groups and insert them to dynamoDB

Test Plan
---------

1. test local ``` python3 -m hmalib.lambdas.fetcher ```, new privacy groups should be added in wangmengyang-HMAConfig table
2. ```black . ```
3. ```mypy hmalib``` make sure no errors

### CODE REVIEW

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:3]: Nice job adding this doc block. very clear

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    collab_config_table = os.environ["THREAT_EXCHANGE_CONFIG_DYNAMODB"]
+    api_key = AWSSecrets.te_api_key()
+    api = ThreatExchangeAPI(api_key)
+    privacy_group_member_list = api.get_threat_privacy_groups_member()
+    privacy_group_owner_list = api.get_threat_privacy_groups_owner()
+    items = []
+    privacy_ids = set()
+    build_items(items, privacy_group_member_list, privacy_ids)
+    build_items(items, privacy_group_owner_list, privacy_ids)
+    table = dynamodb.Table(collab_config_table)
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:35]: What happens if I call this function twice? Will every privacy group be added twice?

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    with table.batch_writer() as batch:
+        for item in items:
+            batch.put_item(Item=item)
+
+
+def build_items(items, privacy_group_list, privacy_ids):
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:41]: Can we add typing here? I'm surprised `mypy` didn't ask you to

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    if privacy_group_list:
+        for privacy_group in privacy_group_list:
+            if (
+                privacy_group.threat_updates_enabled
+                and privacy_group.id not in privacy_ids
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:46]: So it looks like youre checking if the id already exists to prevent duplicates. This seems round about since sets already prevent duplicate entries.

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    build_items(items, privacy_group_member_list, privacy_ids)
+    build_items(items, privacy_group_owner_list, privacy_ids)
+    table = dynamodb.Table(collab_config_table)
+    with table.batch_writer() as batch:
+        for item in items:
+            batch.put_item(Item=item)
+
+
+def build_items(items, privacy_group_list, privacy_ids):
+    if privacy_group_list:
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:42]: What do we need this check for?

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+def lambda_handler(event, context):
+    collab_config_table = os.environ["THREAT_EXCHANGE_CONFIG_DYNAMODB"]
+    api_key = AWSSecrets.te_api_key()
+    api = ThreatExchangeAPI(api_key)
+    privacy_group_member_list = api.get_threat_privacy_groups_member()
+    privacy_group_owner_list = api.get_threat_privacy_groups_owner()
+    items = []
+    privacy_ids = set()
+    build_items(items, privacy_group_member_list, privacy_ids)
+    build_items(items, privacy_group_owner_list, privacy_ids)
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:34]: I find your structure here a bit hard to follow. I think the standard python way to do this is with for loops or list comprehension

**Review by jeberl — CHANGES_REQUESTED** [review]

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    privacy_group_owner_list = api.get_threat_privacy_groups_owner()
+    unique_privacy_groups = set(privacy_group_member_list + privacy_group_owner_list)
+    table = dynamodb.Table(collab_config_table)
+    with table.batch_writer() as batch:
+        for privacy_group in unique_privacy_groups:
+            if privacy_group.threat_updates_enabled:
+                logger.info("Adding collaboration name %s", privacy_group.name)
+                item = {
+                    "Name": privacy_group.name,
+                    "privacy_group": privacy_group.id,
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:39]: I suggest calling this `privacty_group_id` in the table to distinguish it from `privacy_group`. In other places `privacy_group` is a string derived from the file name of the data for that privacy group.

**Review by eriksendc — CHANGES_REQUESTED** [review]

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py (hunk 7 above)
Comment (mengyangwang) [reply hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py]: yeah I agreed with that, it's changed to privacy_group_id

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    api = ThreatExchangeAPI(api_key)
+    privacy_group_member_list = api.get_threat_privacy_groups_member()
+    privacy_group_owner_list = api.get_threat_privacy_groups_owner()
+    unique_privacy_groups = set(privacy_group_member_list + privacy_group_owner_list)
+    table = dynamodb.Table(collab_config_table)
+    with table.batch_writer() as batch:
+        for privacy_group in unique_privacy_groups:
+            if privacy_group.threat_updates_enabled:
+                logger.info("Adding collaboration name %s", privacy_group.name)
+                item = {
```
Comment (schatten) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:37]: You've already noted this, but this would be better migrated to the config framework...

File: hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                    "Name": privacy_group.name,
+                    "privacy_group_id": privacy_group.id,
+                    "tags": [],
```
Comment (schatten) [inline hasher-matcher-actioner/hmalib/lambdas/add_privacy_groups.py:40]: Are the tags always supposed to be empty?

File: hasher-matcher-actioner/hmalib/lambdas/fetcher.py [hunk 10]
```diff
@@ -67,7 +67,7 @@ def lambda_handler(event, context):
     response_iterator = paginator.paginate(
```
Comment (schatten) [inline hasher-matcher-actioner/hmalib/lambdas/fetcher.py:67]: nit: do we need to paginate? dynamodb will happily return 1MB of data. Are we expecting the list of configs to be that long, or are we being defensive?

**Review by schatten — COMMENTED** [review]
Have questions about the idempotency of the lambda. See inline comments for more.

File: hasher-matcher-actioner/hmalib/lambdas/fetcher.py (hunk 10 above)
Comment (mengyangwang) [reply hasher-matcher-actioner/hmalib/lambdas/fetcher.py:102]: yes this part will be changed using config framework in a follow-up pull request

File: hasher-matcher-actioner/hmalib/lambdas/fetcher.py [hunk 11]
```diff
@@ -170,6 +171,33 @@ def lambda_handler(event, context):
...
+    privacy_group_member_list = api.get_threat_privacy_groups_member()
+    privacy_group_owner_list = api.get_threat_privacy_groups_owner()
+    unique_privacy_groups = set(privacy_group_member_list + privacy_group_owner_list)
+
+    for privacy_group in unique_privacy_groups:
+        if privacy_group.threat_updates_enabled:
+            logger.info("Adding collaboration name %s", privacy_group.name)
+            config = ThreatExchangeConfig(
+                privacy_group.name,
+                fetcher_active=True,
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/fetcher.py:195]: I would put a TODO here reminding us to switch the default here to `False` before v0 launch

File: hasher-matcher-actioner/hmalib/lambdas/fetcher.py [hunk 12]
```diff
@@ -170,6 +171,33 @@ def lambda_handler(event, context):
...
+    """
+    lambda_init_once()
+    api_key = AWSSecrets.te_api_key()
+    api = ThreatExchangeAPI(api_key)
+    privacy_group_member_list = api.get_threat_privacy_groups_member()
+    privacy_group_owner_list = api.get_threat_privacy_groups_owner()
+    unique_privacy_groups = set(privacy_group_member_list + privacy_group_owner_list)
+
+    for privacy_group in unique_privacy_groups:
+        if privacy_group.threat_updates_enabled:
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/fetcher.py:191]: maybe add a comment here like "HMA can only read from privacy groups that have threat_updates enabled. See here for more details: https://developers.facebook.com/docs/threat-exchange/reference/apis/threat-updates/v9.0"

**Review by jeberl — CHANGES_REQUESTED** [review]
This is looking great. Mostly small, non-blocking changes except for the one about names. If you really want to ship ASAP you can just promise to fix in a later PR

**Review by jeberl — APPROVED** [review]
Ship it!
